src/librustc_errors/lib.rs

```diff
@@ -322,6 +322,18 @@ pub struct Handler {
     /// this handler. These hashes is used to avoid emitting the same error
     /// twice.
     emitted_diagnostics: Lock<FxHashSet<u128>>,
+
+    /// Stashed diagnostics emitted in one stage of the compiler that may be
+    /// stolen by other stages (e.g. to improve them and add more information).
+    /// The stashed diagnostics count towards the total error count.
+    /// When `.abort_if_errors()` is called, these are also emitted.
+    stashed_diagnostics: Lock<FxIndexMap<(Span, StashKey), Diagnostic>>,
+}
+
+/// A key denoting where from a diagnostic was stashed.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum StashKey {
+    ItemNoType,
```

Comment:
I considered making this a `&'static str` or something but that felt less good... can always change to that in the future or something.

Reply:
I think this is the right approach.